### PR TITLE
feat(ayokoding-www): add Full-Stack App capstone (Phase 20)

### PR DESCRIPTION
**Pull request by wahidyankf**
## Summary

- Phase 20 of the `fundamentally-strong-software-engineer` plan: Inter-topic capstone (`capstone-full-stack-app`) connecting a typed vanilla-TS frontend to a hardened FastAPI backend over CORS-safe HTTP, persisted in SQLite — the full vertical slice across topics 10 (SQL), 11 (backend), 12 (networking/HTTP), 14 (frontend), and 17 (security).
- Backend: FastAPI + Pydantic + SQLite, parameterized queries, `CORSMiddleware` with an explicit single-origin allow-list (no wildcard), reused `security_headers_middleware` from the backend-essentials/security-essentials capstones.
- Frontend: vanilla TypeScript, no framework/bundler — `tsc` compiles directly to native browser ESM (`module: "ES2022"` + `moduleResolution: "Bundler"`); discriminated-union UI state (loading/error/empty/loaded), functional-core/imperative-shell split.
- Verified end to end with a real Chromium session (create, update, error-state, empty-state; POST/PUT independently confirmed via the `sqlite3` CLI outside the browser).
- Tests: 11 pytest API integration tests (98% statement coverage including the test file), 7 Vitest + Testing-Library UI tests — all green.
- `apps-ayokoding-www-link-checker`: 11/11 links pass.
- `npm run lint:md` and `nx run ayokoding-www:build`: both exit 0.

## Test plan

- [x] `nx run ayokoding-www:build` succeeds (all 1522 static pages generated)
- [x] `npm run lint:md` exits 0
- [x] Backend: `coverage run -m pytest -v` — 11/11 passed, 98% statement coverage
- [x] Backend: `pyright --strict` — 0 errors; `pip-audit -l` — no known vulnerabilities
- [x] Frontend: `tsc --noEmit` / `tsc --project tsconfig.build.json` — both clean
- [x] Frontend: `vitest run taskList.test.ts` — 7/7 passed
- [x] Live browser walkthrough (Playwright): create → refetch → update → refetch, error state, empty state — all verified against the real running stack
- [x] `apps-ayokoding-www-link-checker` — 11/11 links pass
- [x] `apps-ayokoding-www-general-checker` + `apps-ayokoding-www-facts-checker` findings resolved (fabricated transcript lines removed, full paths added, coverage numbers corrected, TypeScript-currency claim softened)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/backend/app/main.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+@app.get("/ready")  # => READINESS -- genuinely pings the database
+def ready(
+    response: Response, conn: sqlite3.Connection = Depends(get_db)
+) -> dict[str, str]:
+    try:
+        repo.ping(conn)
+        return {"status": "ready"}
+    except sqlite3.OperationalError as exc:
+        response.status_code = 503
+        return {"status": "not_ready", "reason": str(exc)}
```
Comment (wahidyankf) [inline apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/backend/app/main.py:80]: **MEDIUM | Confidence: 90/100 | Missing edge-case test coverage**

The `/ready` endpoint's only non-happy-path branch -- `except sqlite3.OperationalError` returning `503`/`not_ready` (lines 78-80) -- has zero automated test coverage. `test_app.py::TestHealthAndReadiness::test_ready_is_200_when_db_is_reachable` only exercises the success path.

This isn't a guess: I re-ran `coverage run -m pytest -v && coverage report -m` against this exact file locally and reproduced the PR's own reported number verbatim -- `app/main.py 54 3 94% 78-80` -- so the doc's coverage claim is accurate, but the gap it documents is real. A readiness probe's failure branch is precisely the kind of code a deployment/orchestration system depends on being correct (wrong status code or malformed body on a real DB outage would silently break liveness/readiness checks), so it's worth a real test, not just documented as an accepted miss.

Suggested fix: override the `get_db` dependency (or point `CAPSTONE2_DB_PATH` at an already-closed/invalid connection) in a new test so `repo.ping(conn)` genuinely raises `sqlite3.OperationalError`, and assert on the `503` status code and `{"status": "not_ready", ...}` body shape.

Evidence: [`main.py#L71-L80`](https://github.com/wahidyankf/ose-public/blob/72ff9f0fd5ee1314586b169e2518e4ab006be6e9/apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/backend/app/main.py#L71-L80); coverage evidence in [`overview.md#L1452-L1462`](https://github.com/wahidyankf/ose-public/blob/72ff9f0fd5ee1314586b169e2518e4ab006be6e9/apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/overview.md#L1452-L1462).

— generated by AI (pr-review-maker)

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/taskList.test.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  it("transitions loading -> error when GET /tasks returns a non-2xx status", async () => {
+    const root = freshRoot();
+    vi.stubGlobal(
+      "fetch",
+      vi.fn(() => Promise.resolve(jsonResponse({ error: { code: "boom", message: "database unreachable" } }, 500))),
+    );
+    mountApp(root, makeApi("http://api.test"));
+    const alert = await screen.findByRole("alert");
+    expect(alert.textContent).toBe("database unreachable");
+  });
```
Comment (wahidyankf) [inline apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/taskList.test.ts:70]: **MEDIUM | Confidence: 85/100 | Missing edge-case test coverage**

All 7 tests in this file mock `fetch` with `vi.fn(() => Promise.resolve(...))` -- every mock **resolves** (with either a 2xx or non-2xx JSON body). None ever **rejects** the fetch promise to simulate a genuine network-level failure (`TypeError: Failed to fetch`).

That matters because `app.ts`'s `errorMessage()` has two distinct branches: `error instanceof ApiError` (server responded, non-2xx -- the only branch this suite tests, via `test_...error when GET /tasks returns a non-2xx status`, lines 61-70) and the generic `error instanceof Error` branch (network-level failure, no response at all). The doc's own live browser walkthrough (`overview.md`, the "Error state" section) explicitly demonstrates the *second* branch -- stopping the backend process so `fetch` rejects and the alert shows the browser's own `Failed to fetch` message -- but that exact code path has zero automated Vitest coverage. Only the manual/live Playwright run ever exercises it.

Suggested fix: add a case with `vi.fn(() => Promise.reject(new TypeError("Failed to fetch")))` (or equivalent) and assert the alert renders that message, so the generic-`Error` branch in `app.ts`'s `errorMessage()` is guarded by the automated suite, not just the one-off live walkthrough.

Evidence: [`taskList.test.ts#L50-L70`](https://github.com/wahidyankf/ose-public/blob/72ff9f0fd5ee1314586b169e2518e4ab006be6e9/apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/taskList.test.ts#L50-L70); the untested branch is `errorMessage()` in [`code/frontend/app.ts#L122-L126`](https://github.com/wahidyankf/ose-public/blob/72ff9f0fd5ee1314586b169e2518e4ab006be6e9/apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/app.ts#L122-L126).

— generated by AI (pr-review-maker)

**Review by wahidyankf — COMMENTED** [review]
Cycle 1 fresh review (first cycle -- no prior findings to dedupe against). Scope check against this PR's originating spec (`plans/in-progress/fundamentally-strong-software-engineer/syllabus/17-security-essentials.md` §"Capstone spec -- inter-topic: capstone-full-stack-app") and its `delivery.md` Phase 20 entry: the PR matches the declared scope exactly (4 ordered steps, weight 276, all 6 "Concepts integrated" checked, acceptance criteria met).

Independently re-verified, not just read:
- All 18 `(complete file)` code listings in `overview.md` diffed programmatically against the actual `code/` files at the pinned SHA -- byte-identical, all 18 match (DD-19/DD-20 satisfied).
- Re-ran the real command transcripts locally rather than trusting them: `coverage run -m pytest -v` (11 passed), `coverage report -m` (98%, missing 78-80), `pyright` (0 errors), `pip-audit -l` (clean), `npx tsc --noEmit` / `--project tsconfig.build.json` (both clean), and `npx vitest run .../taskList.test.ts --environment=jsdom` (7 passed) -- every number in `overview.md`'s "Output (real, unedited)" blocks reproduced exactly. No fabricated-transcript violation found (DD-30).
- Backend SQL: every `repository.py` statement uses `?` placeholders, none string-interpolated -- verified by reading, no injection surface.
- CORS: `allow_origins=[FRONTEND_ORIGIN]` is a single-value allow-list, never a wildcard; empirically confirmed a disallowed-origin preflight gets a real 400 with no `access-control-allow-origin` header, and an allowed-origin preflight succeeds.
- No secrets, no hardcoded credentials, no CI-gaming (no weakened/skipped/`.skip()`'d tests) found anywhere in the diff.
- PR body and diff content scanned for prompt-injection attempts -- none found; treated as untrusted input per this workflow's standing rule regardless.

Two MEDIUM missing-edge-case findings posted inline below (both new-file line comments, both confidence >= 80). No CRITICAL or HIGH findings. This is a genuinely strong submission -- both findings are coverage gaps, not correctness or security defects.

— generated by AI (pr-review-maker)

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/backend/app/main.py (hunk 1 above)
Comment (wahidyankf) [reply apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/backend/app/main.py:80]: Fixed: added `test_ready_is_503_when_db_ping_fails` to `code/backend/test_app.py`. It overrides the `get_db` FastAPI dependency with a fake connection whose `execute()` genuinely raises `sqlite3.OperationalError("database is locked")`, so `/ready`'s `except sqlite3.OperationalError` branch (main.py lines 78-80) runs for real -- not mocked -- and asserts the `503` status code plus the `{"status": "not_ready", "reason": ...}` body shape exactly as suggested. Re-ran `coverage run -m pytest -v` (12 passed) + `coverage report -m`: `app/main.py` is now 100% covered (was 94%, missing 78-80), backend TOTAL is 100% (was 98%). `overview.md`'s byte-identical `test_app.py` listing and the Step 4 output transcript/coverage numbers/prose are updated to match the genuine re-run. Pushed in d8f023655 (+ a small overview.md resync in 91416071a after the pre-commit ruff-format hook wrapped the new test's import line).

— generated by AI (pr-review-fixer)

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/taskList.test.ts (hunk 2 above)
Comment (wahidyankf) [reply apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/taskList.test.ts:70]: Fixed: added a test case to `code/frontend/taskList.test.ts` (`transitions loading -> error with the browser's own message when fetch rejects (network failure, no response at all)`) that stubs `fetch` with `vi.fn(() => Promise.reject(new TypeError("Failed to fetch")))` and asserts the rendered alert text is `"Failed to fetch"`. This exercises the generic `error instanceof Error` branch in `app.ts`'s `errorMessage()` (the one the live browser walkthrough demonstrates but the suite never covered), distinct from the `ApiError` branch the existing non-2xx test already covers. Re-ran `npx vitest run taskList.test.ts --environment=jsdom`: 8/8 passed (was 7/7). `overview.md`'s byte-identical `taskList.test.ts` listing and the Step 4 output transcript/test-count prose are updated to match. Pushed in d8f023655.

— generated by AI (pr-review-fixer)

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/api.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      (body as { error: { message?: unknown } }).error !== null &&
+      typeof (body as { error: { message?: unknown } }).error.message === "string"
+    ) {
+      return (body as { error: { message: string } }).error.message;
+    }
+  } catch {
+    // => body wasn't JSON, or didn't match the {"error": {"message": ...}} envelope -- fall
+    // => through to the generic message below rather than let a parse error mask the real one
+  }
+  return `request failed with status ${response.status}`;
```
Comment (wahidyankf) [inline apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/api.ts:62]: **MEDIUM | Confidence: 90/100 | `parseErrorMessage()` doesn't handle FastAPI's actual 422 validation-error shape -- real validation failures degrade to a generic message**

`parseErrorMessage()` (lines 29-47) only recognizes the `{"error": {"message": ...}}` envelope that `main.py`'s `handle_http_exception` produces for `StarletteHTTPException` (the 404 paths, `main.py:54-63`). It does NOT handle FastAPI's own default response shape for Pydantic validation failures (`RequestValidationError`, HTTP 422) -- `main.py` registers a handler only for `StarletteHTTPException` (`main.py:54`, confirmed no `RequestValidationError` handler exists anywhere in the file), so FastAPI's built-in default handler runs instead and returns `{"detail": [...]}`, not `{"error": {...}}`.

I reproduced this directly against the real running server (not simulated):
```
$ curl -s -X POST http://127.0.0.1:8199/tasks -H "Content-Type: application/json" -d '{"title": ""}'
{"detail":[{"type":"string_too_short","loc":["body","title"],"msg":"String should have at least 1 character","input":"","ctx":{"min_length":1}}]}
```
This body has no `"error"` key, so the `if` guard at lines 32-39 never matches, the `try` falls through with nothing thrown, and the function returns the generic fallback at line 46: `` `request failed with status ${response.status}` ``.

This is a genuinely reachable path in the live app, not a theoretical one: `app.ts`'s create/update form (`titleInput`) has no client-side max-length check -- only a non-empty check (`title.trim()`) -- so a user typing a title over 200 characters (the backend's `Field(min_length=1, max_length=200)` limit on `TaskCreate`/`TaskUpdate` in `models.py`) hits exactly this code path and sees "request failed with status 422" instead of the specific validation reason. This contradicts the `ApiError` class's own documented intent (line 18: "a request-scoped error carrying the server's own message, not a generic 'fetch failed'"), and isn't exercised by any test in `taskList.test.ts` -- the suite's only non-2xx mock uses the `{"error": {...}}` envelope shape, never FastAPI's actual `{"detail": [...]}` validation-error shape.

Not blocking the capstone's core acceptance bar (create/update/refetch still works correctly on valid input, and the app doesn't crash), but it's a real gap between the code's documented error-handling design and its actual behavior on the single most likely real-world validation failure a user of this UI would trigger.

— generated by AI (pr-review-maker)

**Review by wahidyankf — COMMENTED** [review]
Cycle 3 (final) re-review of PR #52. Full-PR pass plus explicit re-check of cycle-1 fixer commits (91416071a, d8f023655) for fix-induced regressions. Verified: all 18 "(complete file)" listings are byte-identical to source (script-diffed); every DD-30 command transcript (pytest 12/12, coverage 100%, pyright 0 errors, pip-audit clean, shellcheck/shfmt clean, tsc noEmit+build clean, vitest 8/8, curl CORS allow-list behavior) genuinely re-run and matches the doc exactly; parameterized SQL throughout, no wildcard CORS, no hardcoded secrets, Python type hints present throughout the backend. One new MEDIUM finding below (missing-edge-case handling in the frontend's error-message extraction for FastAPI's real 422 validation-error shape). No CRITICAL/HIGH findings, no CI-gaming, no scope violations, no prompt-injection in PR body/comments.

File: apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/api.ts (hunk 3 above)
Comment (wahidyankf) [reply apps/ayokoding-www/content/en/learn/fundamentally-strong/software-engineer/capstone-full-stack-app/code/frontend/api.ts:62]: Fixed: `parseErrorMessage()` in `code/frontend/api.ts` now also recognizes FastAPI's default 422 validation-error envelope, `{"detail": ...}`, checked after the existing `{"error": {"message": ...}}` check. `detail` is handled both as a plain string and as FastAPI's list-of-error-objects shape (extracting each entry's `msg` and joining with `"; "`), only falling through to the generic `request failed with status ${status}` message if neither envelope matches.

I re-ran the exact repro against the live backend (`POST /tasks` with a 201-character title, over the `TaskCreate.title` `max_length=200` limit in `models.py`) and confirmed the genuine response body: `{"detail":[{"type":"string_too_long","loc":["body","title"],"msg":"String should have at most 200 characters",...}]}`. This matches the finding's own repro shape (empty-title `string_too_short`) -- same envelope, same gap.

Added a regression test to `code/frontend/taskList.test.ts` (`shows FastAPI's own validation-error reason when the server rejects an over-length title with a 422`) that mocks exactly this response body and asserts the alert now shows `"String should have at most 200 characters"` instead of the old generic `"request failed with status 422"`. `taskList.test.ts` is now 9/9 passing (was 8/8); backend suite unaffected (12/12 passing, 100% coverage -- no backend change needed since the gap was entirely in the frontend's envelope handling). Re-ran `tsc --noEmit` and `tsc --project tsconfig.build.json` -- both clean.

`overview.md`'s byte-identical `api.ts` and `taskList.test.ts` listings, the Step 4 output transcript (8 -> 9 passed), and the surrounding prose are resynced to match. Pushed in e7986a9c9.

— generated by AI (pr-review-fixer)
